File: sdk/python/src/hudson_sandbox/_stream.py
from __future__ import annotations
import asyncio
import base64
import dataclasses
import httpx
from ._wire import ClientError, decode_model, parse_json


@dataclasses.dataclass(repr=False)
class Event:
    event: str
    data: object
    cursor: str | None = None

    def __repr__(self):
        return "Event(...)"

    def to_wire(self):
        value = {"event": self.event, "data": self.data.to_wire()}
        if self.cursor is not None:
            value["cursor"] = self.cursor
        return value


class Parser:
    def __init__(self):
        self.line = bytearray()
        self.data = []
        self.event = ""
        self.id = None
        self.frame_bytes = 0
        self.after_cr = False
        self.first = True

    def byte(self, byte):
        if self.after_cr and byte == 10:
            self.after_cr = False
            return None
        self.after_cr = byte == 13
        self.frame_bytes += 1
        if self.frame_bytes > 65536:
            raise ClientError("protocol")
        if byte not in (10, 13):
            self.line.append(byte)
            return None
        try:
            line = self.line.decode("utf-8")
        except UnicodeError:
            raise ClientError("protocol") from None
        self.line.clear()
        if self.first:
            line = line.lstrip("\ufeff")
            self.first = False
        if not line:
            self.frame_bytes = 0
            data, kind, cursor = "\n".join(self.data), self.event, self.id
            self.data = []
            self.event = ""
            self.id = None
            if not data:
                return None
            return decode_event(kind, cursor, data)
        if line.startswith(":"):
            return None
        field, _, value = line.partition(":")
        if value.startswith(" "):
            value = value[1:]
        if field == "event":
            self.event = value
        elif field == "id" and "\0" not in value:
            self.id = value
        elif field == "data":
            self.data.append(value)
        return None


def decode_event(kind, cursor, raw):
    if kind in ("output", "end"):
        if (
            not cursor
            or len(cursor.encode()) > 2048
            or any(ord(c) < 32 or 127 <= ord(c) <= 159 for c in cursor)
        ):
            raise ClientError("protocol")
    elif kind not in ("gap", "error") or cursor is not None:
        raise ClientError("protocol")
    name = {
        "output": "OutputEvent",
        "end": "EndEvent",
        "gap": "StreamProblemEvent",
        "error": "StreamProblemEvent",
    }[kind]
    data = decode_model(name, parse_json(raw))
    if kind == "output":
        try:
            binary = base64.b64decode(data.data_base64, validate=True)
        except ValueError:
            raise ClientError("protocol") from None
        if (
            base64.b64encode(binary).decode() != data.data_base64
            or len(binary) > 32768
            or data.stream not in ("stdout", "stderr")
            or data.offset + len(binary) != data.next_offset
            or data.next_offset > data.stored
            or data.stored > 10485760
            or data.stored > data.seen
            or data.at_end != (data.next_offset == data.stored)
            or not data.guest_reported
            or data.truncated != (data.seen > data.stored)
        ):
            raise ClientError("protocol")
    if kind == "end":
        if (
            data.reason != "complete"
            or not data.guest_reported
            or any(
                s.stored > s.seen
                or s.stored > 10485760
                or s.truncated != (s.seen > s.stored)
                for s in (data.stdout, data.stderr)
            )
        ):
            raise ClientError("protocol")
    return Event(kind, data, cursor)


class EventStream:
    def __init__(self, response, deadline):
        self._response = response
        self._deadline = deadline
        self._chunks = response.aiter_raw().__aiter__()
        self._pending = b""
        self._position = 0
        self._parser = Parser()
        self._ended = False

    def __repr__(self):
        return "EventStream(...)"

    def __aiter__(self):
        return self

    async def __aenter__(self):
        return self

    async def __aexit__(self, *_):
        await self.aclose()

    async def aclose(self):
        self._ended = True
        await self._response.aclose()

    async def __anext__(self):
        if self._ended:
            raise StopAsyncIteration
        try:
            if asyncio.get_running_loop().time() >= self._deadline:
                raise ClientError("transport")
            async with asyncio.timeout_at(self._deadline):
                while True:
                    while self._position < len(self._pending):
                        byte = self._pending[self._position]
                        self._position += 1
                        event = self._parser.byte(byte)
                        if event is not None:
                            if event.event != "output":
                                await self.aclose()
                            return event
                    try:
                        self._pending = await anext(self._chunks)
                    except StopAsyncIteration:
                        await self.aclose()
                        raise  # Drop incomplete frames; never advance a cursor.
                    self._position = 0
                    if len(self._pending) > 262144:
                        raise ClientError("protocol")
        except (httpx.HTTPError, TimeoutError):
            await self.aclose()
            raise ClientError("transport") from None
        except BaseException:
            await self.aclose()
            raise

File: sdk/typescript/src/stream.ts
import { inspect } from 'node:util';
import type { Response } from './transport.js';
import type { OutputEvent, EndEvent, StreamProblemEvent } from './models.js';
import { ClientError, fail, model, parseJson } from './wire.js';
export type Event = { event: 'output'; data: OutputEvent; cursor: string } | { event: 'end'; data: EndEvent; cursor: string } | { event: 'gap' | 'error'; data: StreamProblemEvent };
export function decodeEvent(kind: string, cursor: string | undefined, raw: string): Event {
  if (kind === 'output' || kind === 'end') {
    if (!cursor || Buffer.byteLength(cursor) > 2048 || /[\x00-\x1f\x7f-\x9f]/u.test(cursor)) fail('protocol');
  } else if ((kind !== 'gap' && kind !== 'error') || cursor !== undefined) fail('protocol');
  if (kind === 'output') {
    const data = model('OutputEvent', parseJson(raw)) as unknown as OutputEvent;
    const bytes = Buffer.from(data.data_base64, 'base64');
    if (bytes.toString('base64') !== data.data_base64 || bytes.length > 32768 || !['stdout','stderr'].includes(data.stream) || data.offset + BigInt(bytes.length) !== data.next_offset || data.next_offset > data.stored || data.stored > 10485760n || data.stored > data.seen || data.at_end !== (data.next_offset === data.stored) || !data.guest_reported || data.truncated !== (data.seen > data.stored)) fail('protocol');
    return { event: kind, data, cursor: cursor! };
  }
  if (kind === 'end') {
    const data = model('EndEvent', parseJson(raw)) as unknown as EndEvent;
    if (data.reason !== 'complete' || !data.guest_reported || [data.stdout, data.stderr].some(s => s.stored > s.seen || s.stored > 10485760n || s.truncated !== (s.seen > s.stored))) fail('protocol');
    return { event: kind, data, cursor: cursor! };
  }
  return { event: kind as 'gap' | 'error', data: model('StreamProblemEvent', parseJson(raw)) as unknown as StreamProblemEvent };
}
export class Parser {
  #line: number[] = []; #data: string[] = []; #event = ''; #id?: string;
  #bytes = 0; #afterCr = false; #first = true;
  byte(byte: number): Event | undefined {
    if (this.#afterCr && byte === 10) { this.#afterCr = false; return; }
    this.#afterCr = byte === 13; if (++this.#bytes > 65536) fail('protocol');
    if (byte !== 10 && byte !== 13) { this.#line.push(byte); return; }
    let line: string;
    try { line = new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(Uint8Array.from(this.#line)); } catch { return fail('protocol'); }
    this.#line = [];
    if (this.#first) { if (line.startsWith('\ufeff')) line = line.slice(1); this.#first = false; }
    if (!line) {
      this.#bytes = 0; const data = this.#data.join('\n'), kind = this.#event, cursor = this.#id;
      this.#data = []; this.#event = ''; this.#id = undefined;
      return data ? decodeEvent(kind, cursor, data) : undefined;
    }
    if (line.startsWith(':')) return;
    const colon = line.indexOf(':'); const field = colon < 0 ? line : line.slice(0, colon);
    let value = colon < 0 ? '' : line.slice(colon + 1); if (value.startsWith(' ')) value = value.slice(1);
    if (field === 'event') this.#event = value;
    else if (field === 'id' && !value.includes('\0')) this.#id = value;
    else if (field === 'data') this.#data.push(value);
  }
}
export class EventStream implements AsyncIterableIterator<Event> {
  #response: Response; #chunks: AsyncIterator<Uint8Array>; #pending = new Uint8Array(); #position = 0; #parser = new Parser(); #ended = false;
  constructor(response: Response) { this.#response = response; this.#chunks = response.message[Symbol.asyncIterator](); }
  [inspect.custom](): string { return 'EventStream(...)'; }
  [Symbol.asyncIterator](): AsyncIterableIterator<Event> { return this; }
  close(): void { this.#ended = true; this.#response.close(); }
  async [Symbol.asyncDispose](): Promise<void> { this.close(); }
  async return(): Promise<IteratorResult<Event>> { this.close(); return { done: true, value: undefined }; }
  async next(): Promise<IteratorResult<Event>> {
    if (this.#ended) return { done: true, value: undefined };
    try {
      while (true) {
        if (performance.now() >= this.#response.deadline) fail('transport');
        while (this.#position < this.#pending.length) {
          const event = this.#parser.byte(this.#pending[this.#position++]!);
          if (event) { if (event.event !== 'output') this.close(); return { done: false, value: event }; }
        }
        const next = await this.#chunks.next();
        if (next.done) { this.close(); return { done: true, value: undefined }; }
        if (next.value.byteLength > 262144) fail('protocol'); this.#pending = new Uint8Array(next.value); this.#position = 0;
      }
    } catch (error) { this.close(); if (error instanceof ClientError) throw error; throw new ClientError('transport'); }
  }
}
